lib/Sema/TypeCheckDeclOverride.cpp

```diff
@@ -928,6 +931,96 @@ static void checkOverrideAccessControl(ValueDecl *baseDecl, ValueDecl *decl,
   }
 }
 
+static GenericSignature *getNewGenericSignature(ValueDecl *base,
+                                                ValueDecl *derived) {
+  auto baseGenericCtx = base->getAsGenericContext();
+  auto derivedGenericCtx = derived->getAsGenericContext();
+  auto &ctx = base->getASTContext();
+
+  if (!baseGenericCtx) {
+    return nullptr;
+  }
+
+  auto baseClass =
+      baseGenericCtx->getInnermostTypeContext()->getSelfClassDecl();
+  auto derivedClass =
+      derivedGenericCtx->getInnermostTypeContext()->getSelfClassDecl();
+
+  if (!derivedClass) {
+    return nullptr;
+  }
+
+  if (!derivedClass->getInterfaceType()->is<LValueType>()) {
+    return nullptr;
+  }
+
+  auto subMap = derivedClass->getInterfaceType()->getContextSubstitutionMap(
+      derivedClass->getModuleContext(), baseClass);
+
+  auto *genericParams = baseGenericCtx->getGenericParams();
+  if (genericParams) {
```

Comment:
I think the condition you care about is that the base method is in a generic context, not that the base class has its own generic parameters. Eg,

```
class Outer<T> {
  class Inner {
    func method()
  }
}
```

Reply:
@slavapestov Do you mean something like `base->getDeclContext()->isGenericContext()`? Although that will fail for all base classes that aren't generic (but the method is) and thus this function would bail out and return nullptr.

```swift
protocol SR_4206_P0 {}

protocol SR_4206_P1: SR_4206_P0 {}
protocol SR_4206_P2: SR_4206_P0 {}

/*class SR_4206_C1 {
  func foo<T: SR_4206_P1>(param: T) {} // expected-note {{overridden declaration is here}}
}

class SR_4206_C2: SR_4206_C1 {
  override func foo<T: SR_4206_P2>(param: T) {} // expected-error {{cannot override method 'foo' with a different generic signature}}
}*/
```

The method isn't in a generic context here for example.

Reply:
No, you want `baseGenericCtx->getGenericSignature() != nullptr`, precisely because the method might have its own generic signature.

Reply:
The generic signature being built here should consist of all the generic parameters of the base class, plus those from the method that are not also from the derived class. So the list of parameters built should never be totally empty.